File: src/main/java/pw/vodes/rimuru/command/commands/CommandBan.java
package pw.vodes.rimuru.command.commands;

import java.time.Instant;
import java.util.ArrayList;

import org.apache.commons.lang3.StringUtils;
import org.javacord.api.entity.permission.Role;
import org.javacord.api.entity.user.User;
import org.javacord.api.event.message.MessageCreateEvent;

import pw.vodes.rimuru.Main;
import pw.vodes.rimuru.audit.AuditLogs;
import pw.vodes.rimuru.audit.CommandStaffAction;
import pw.vodes.rimuru.audit.StaffActionType;
import pw.vodes.rimuru.command.Command;
import pw.vodes.rimuru.command.CommandType;

public class CommandBan extends Command {

	public CommandBan() {
		super("Ban", new String[] {"ban", "b"}, CommandType.mod);
		setUsage("`!ban <user_id/mentioned users> \"reason for ban\"`");
	}

	@Override
	public void run(MessageCreateEvent event) {
		var users = new ArrayList<User>();
		var args = getSplitMessage(event);
		
		if(!event.getMessage().getMentionedUsers().isEmpty()) {
			for(var u : event.getMessage().getMentionedUsers()) {
				var isAdminOrMod = Main.getServer().isAdmin(u);
				for(var role : Main.getConfig().getModRoles()) {
					if(role.hasUser(u))
						isAdminOrMod = true;
				}
				
				if(isAdminOrMod)
					continue;
				users.add(u);
			}
		} else if(StringUtils.isNumeric(args.get(1))) {
			try {
				var u = Main.api.getUserById(args.get(1)).get();
				if(Main.getServer().isMember(u)) {
					var isAdminOrMod = Main.getServer().isAdmin(u);
					for(var role : Main.getConfig().getModRoles()) {
						if(role.hasUser(u))
							isAdminOrMod = true;
					}
					if(!isAdminOrMod) {
						users.add(u);
					}
				} else {
					users.add(u);
				}
			} catch (Exception e) { e.printStackTrace(); }
		} else {
			event.getChannel().sendMessage(getUsage());
			return;
		}
		
		var reason = "";
		
		if(users.size() < 2 && !args.get(2).isBlank()) {
			reason = args.get(2);
		}
		
		for(var u : users) {
			AuditLogs.registerStaffAction(new CommandStaffAction(Instant.now().getEpochSecond()
			, u.getIdAsString()
			, event.getMessageAuthor().getIdAsString()
			, reason
			, StaffActionType.ban));
			Main.getServer().banUser(u);
		}
		event.getMessage().delete();
	}

}

File: src/main/kotlin/pw/vodes/rimurukt/command/commands/CommandBan.kt
package pw.vodes.rimurukt.command.commands

import org.javacord.api.event.message.MessageCreateEvent
import org.javacord.api.interaction.SlashCommandBuilder
import org.javacord.api.interaction.SlashCommandInteraction
import org.javacord.api.interaction.SlashCommandOption
import pw.vodes.rimurukt.Main
import pw.vodes.rimurukt.command.Command
import pw.vodes.rimurukt.command.CommandType
import pw.vodes.rimurukt.deleteAfter
import pw.vodes.rimurukt.epochSecond
import pw.vodes.rimurukt.reply
import pw.vodes.rimurukt.services.AuditLogs
import pw.vodes.rimurukt.services.StaffAction
import pw.vodes.rimurukt.services.StaffActionType
import java.util.concurrent.TimeUnit
import kotlin.jvm.optionals.getOrNull

class CommandBan : Command("Ban", arrayOf("ban"), CommandType.MOD, "ban") {

    init {
        usage = "`${Main.config.commandPrefix}ban <user_id/mentioned users> \"reason for ban\"`"
    }

    override fun run(event: MessageCreateEvent) {
        val args = args(event).toMutableList()
        if (args[1].isEmpty())
            event.channel.sendMessage(usage).also { return }

        args.removeAt(0)

        val users = listedUsers(event.messageContent)
        if (users.isEmpty())
            event.channel.sendMessage("No valid users were passed.").also { return }

        users.forEach {
            if (!canKickOrBan(event.messageAuthor.asUser().get(), it, event.server.get(), false)) {
                event.channel.sendMessage("Cannot ban ${it.name}.").deleteAfter(5)
            } else {
                event.server.get().banUser(it).thenAccept { _ ->
                    AuditLogs.registerStaffAction(
                        StaffAction(
                            it.idAsString,
                            event.messageAuthor.idAsString,
                            epochSecond(),
                            StaffActionType.BAN,
                            args.find { s -> !s.matches("<?@?(\\d{17,20})>?".toRegex()) } ?: ""
                        )
                    )
                }
            }
        }
        event.message.deleteAfter(1, TimeUnit.SECONDS)
    }

    override fun getSlashCommandBuilder() = SlashCommandBuilder()
        .setName("ban")
        .setDescription("Bans users from the server.")
        .setEnabledInDms(false)
        .addOption(SlashCommandOption.createUserOption("user", "User to ban", true))
        .addOption(SlashCommandOption.createStringOption("reason", "Reason for banning", false))

    override fun runSlashCommand(interaction: SlashCommandInteraction) {
        val target = interaction.arguments[0].userValue.get()
        val server = interaction.server.getOrNull() ?: Main.server

        if (!canKickOrBan(interaction.user, target, server, false))
            interaction.reply("Cannot ban this user.", true).also { return }

        server.banUser(target).thenAccept {
            interaction.reply("User banned.", true)
            AuditLogs.registerStaffAction(
                StaffAction(
                    target.idAsString,
                    interaction.user.idAsString,
                    epochSecond(),
                    StaffActionType.BAN,
                    interaction.arguments.getOrNull(1)?.stringValue?.orElse("") ?: ""
                )
            )
        }.exceptionally {
            interaction.reply("Failed to ban user. Please check the logs.", true)
            null
        }
    }
}